### PR TITLE
Change circulation category in mass

### PR DESCRIPTION
**Pull request by BadrAly**
A new cli set_circulation_category to allow changing the item
circulation category for as many items with one command.

Co-Authored-by: Aly Badr aly.badr@rero.ch

### CODE REVIEW

**Review by iGormilhit — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -1,2 +1,12 @@
 # rero-ils-tools
 RERO ILS misc tools for data manipulation.
+
+export PYTHONPATH=~/rero-ils-tools
```
Comment (rerowep) [inline README.md]: This will  not be correct in a POD.

File: rero_ils_tools/__init__.py [hunk 2]
```diff
@@ -21,7 +21,10 @@
 import click
 from flask.cli import FlaskGroup
 from invenio_app.factory import create_app
-from .cli import app
+
+from .cli.test import app
```
Comment (rerowep) [inline rero_ils_tools/__init__.py]: Maybe test is not the right name for this file?

File: rero_ils_tools/cli/test.py [hunk 3]
```diff
@@ -22,7 +22,7 @@
```
Comment (rerowep) [inline rero_ils_tools/cli/test.py]: Why we need this file?

File: rero_ils_tools/cli/update/circ_category.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        name, ext = os.path.splitext(infile.name)
+        out_file_name = f'{name}_output{ext}'
+        out_file = open(out_file_name, 'w')
+        out_file.write('[\n')
+
+    if save_errors:
+        name, ext = os.path.splitext(infile.name)
+        err_file_name = f'{name}_errors{ext}'
+        error_file = open(err_file_name, 'w')
+        error_file.write('[\n')
```
Comment (rerowep) [inline rero_ils_tools/cli/update/circ_category.py]: There is an ` JsonWriter` class for writing json files.

File: rero_ils_tools/cli/update/circ_category.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+    click.secho(f'Setting circulation category {record_type}', fg='green')
+
+    record_class = get_record_class_from_schema_or_pid_type(
+        pid_type=record_type)
+
+    ids = []
+    for counter, record in enumerate(file_data):
+        record_pid = record.get('pid')
+        new_circ_category = record.get('new_circulation_category')
```
Comment (rerowep) [inline rero_ils_tools/cli/update/circ_category.py]: Not clear if this is an pid?

File: rero_ils_tools/cli/update/circ_category.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+            if save_errors:
+                error_file.write(json.dumps(record, indent=2))
+            continue
+
+        try:
+            if record_type == 'item':
+                record['item_type'] = {
+                    '$ref': get_ref_for_pid('item_types', new_circ_category)
+                }
+                record.update(record, dbcommit=True, reindex=True)
```
Comment (rerowep) [inline rero_ils_tools/cli/update/circ_category.py]: Committing and reindexing every time will be slow for big changes.

File: rero_ils_tools/cli/update/circ_category.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+                ids.append(record.id)
```
Comment (rerowep) [inline rero_ils_tools/cli/update/circ_category.py:117]: This we need only if we bulk index later.

File: rero_ils_tools/cli/update/circ_category.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+                out_file.write(
+                    json.dumps(record, indent=2, separators=(',', ': ')))
+        except Exception as err:
+            text = f'record# {counter} pid {record_pid} failed creation {err}'
+            click.secho(text, fg='red')
+            if save_errors:
+                error_file.write(
+                    json.dumps(record, indent=2, separators=(',', ': ')))
+
+        if counter % 1000 == 0:
```
Comment (rerowep) [inline rero_ils_tools/cli/update/circ_category.py:127]: This we need only if `dbcommit` is false.

**Review by rerowep — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (iGormilhit) [reply README.md]: Maybe the message should precise for which use case it is : is it when one wants to test locally ?

File: README.md [hunk 9]
```diff
@@ -1,2 +1,16 @@
 # rero-ils-tools
 RERO ILS misc tools for data manipulation.
+
+For local machines:
+    export PYTHONPATH=~/rero-ils-tools
+    export PATH=$PATH:~/rero-ils-tools/scripts/
+For PODs:
+    export PYTHONPATH=/invenio/src/rero-ils-tools
+    export PATH=$PATH:/invenio/src/rero-ils-tools/scripts/
+
```
Comment (iGormilhit) [inline README.md]: You can check the rendering on https://github.com/BadrAly/rero-ils-tools/tree/baa-item_type_modify

```suggestion
## Usage

### For local machines

1. Clone the repository.
1. Set the environment variables:

    export PYTHONPATH=~/rero-ils-tools
    export PATH=$PATH:~/rero-ils-tools/scripts/
    
### For PODs

    export PYTHONPATH=/invenio/src/rero-ils-tools
    export PATH=$PATH:/invenio/src/rero-ils-tools/scripts/
```

File: README.md [hunk 10]
```diff
@@ -1,2 +1,16 @@
...
+
+Call_example:
+
+````
+poetry run tools.py tools update set_circulation_category --help
+```
```
Comment (iGormilhit) [inline README.md]:
```suggestion
### Examples

    poetry run tools.py tools update set_circulation_category --help
```

**Review by iGormilhit — COMMENTED** [review]
Improvements on the README.

File: README.md [hunk 11]
```diff
@@ -1,2 +1,20 @@
 # rero-ils-tools
 RERO ILS misc tools for data manipulation.
+
+## Usage
```
Comment (jma) [inline README.md:5]: nice

File: rero_ils_tools/cli/update/circ_category.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+            if save_errors:
+                error_file.write(record)
+        # TODO: create a separate loop for indexing and commits
+        if counter % 1000 == 0:
+            print('DBCOMMIT', counter)
+            db.session.commit()
+            IlsRecordsIndexer().bulk_index(ids, doc_type=record_type)
+            ids = []
+
+    print('DBCOMMIT', counter)
```
Comment (jma) [inline rero_ils_tools/cli/update/circ_category.py]: can be removed

File: rero_ils_tools/cli/update/circ_category.py [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+            if output:
+                out_file.write(new_record)
+        except Exception as err:
+            text = f'record# {counter} pid {record_pid} failed creation {err}'
+            click.secho(text, fg='red')
+            if save_errors:
+                error_file.write(record)
+        # TODO: create a separate loop for indexing and commits
+        if counter % 1000 == 0:
+            print('DBCOMMIT', counter)
```
Comment (jma) [inline rero_ils_tools/cli/update/circ_category.py]: can be removed

File: rero_ils_tools/cli/update/circ_category.py [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+            db.session.commit()
+            IlsRecordsIndexer().bulk_index(ids, doc_type=record_type)
+            ids = []
+
+    print('DBCOMMIT', counter)
+    db.session.commit()
```
Comment (jma) [inline rero_ils_tools/cli/update/circ_category.py]: You can do this only if some ids are left.

**Review by jma — APPROVED** [review]
I would be great to have a commit message with the same rules/format than rero-ils.

**Review by rerowep — APPROVED** [review]

**Pull request merged**